server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/GitRepositoryV2.java

```diff
@@ -0,0 +1,857 @@
+            if (exceedsMinRetention(secondaryRepo, minRetentionCommits, minRetentionDays)) {
+                promoteSecondaryRepo(secondaryRepo);
+            }
+        } else if (exceedsMinRetention(primaryRepo, minRetentionCommits, minRetentionDays)) {
+            createSecondaryRepo();
+        }
+    }
+
+    private static boolean exceedsMinRetention(InternalRepository repo,
+                                               int minRetentionCommits, int minRetentionDays) {
+        final Revision headRevision = repo.headRevision();
+        final Revision firstRevision = repo.firstRevision();
+        if (minRetentionCommits != 0) {
+            if (headRevision.major() - firstRevision.major() <= minRetentionCommits) {
+                return false;
+            }
+        }
+
+        if (minRetentionDays != 0) {
+            final Instant secondCommitCreationTime = repo.secondCommitCreationTimeInstant();
+            return secondCommitCreationTime != null &&
+                   secondCommitCreationTime.isBefore(Instant.now().minus(minRetentionDays, ChronoUnit.DAYS));
+        }
+        return true;
+    }
+
+    private void promoteSecondaryRepo(InternalRepository secondaryRepo) {
+        writeLock();
+        try {
+            logger.info("Promoting the secondary repository in {}/{}.", parent.name(), originalRepoName);
+            repoMetadata.setPrimaryRepoDir(secondaryRepo.jGitRepo().getDirectory());
+            final InternalRepository primaryRepo = this.primaryRepo;
+            this.primaryRepo = secondaryRepo;
+            this.secondaryRepo = null;
+            repositoryWorker.execute(() -> {
+                closeInternalRepository(primaryRepo);
+                final File repoDir = primaryRepo.repoDir();
+                final Path path = repoDir.toPath();
+                final Path newPath = path.resolveSibling(path.getFileName().toString() + SUFFIX_REMOVED);
+                try {
+                    Files.move(path, newPath);
+                } catch (IOException e) {
+                    logger.warn("Failed to mark the old primary repository: {} to {}", repoDir, newPath);
+                }
+            });
+            logger.info("Promotion is done for {}/{}. {} is now the primary.",
+                        parent.name(), originalRepoName, secondaryRepo.repoDir());
+        } finally {
+            writeUnLock();
+        }
+
+        createSecondaryRepo();
+    }
+
+    private void createSecondaryRepo() {
+        InternalRepository secondaryRepo = null;
+        try {
+            writeLock();
+            final Revision headRevision = this.headRevision;
+            isCreatingSecondaryRepo = true;
+            writeUnLock();
+
+            logger.info("Creating the secondary repository in {}/{}. head revision: {}.",
+                        parent.name(), originalRepoName, headRevision);
+            final Map<String, Entry<?>> entries = find(headRevision, ALL_PATH, ImmutableMap.of()).join();
```

Comment:
Should we use multiple of `minRetentionCommits` so that all replias have the same histories?

Reply:
Thanks for the valuable comment. 😄 
I introduced `CreateRollingRepositoryCommand` and made the leader use its own head revision to create the rolling repository. PTAL. 🙇